Add hotkey filtering functionality (by name and by context)

dragunoff opened the pull request:
This implements the UI part of #19898 (back-end part was done in #19900). Hotkey can be filtered by name and by context. Context overlap is used for validation so this new UI should make it clearer to the user how hotkeys are validated. The error message for duplicate bindings is expanded to clarify that.

![image](https://user-images.githubusercontent.com/1355810/163568343-27a1bec8-71af-47e4-8402-8500f1e9e254.png)
_Filtering by context_

![image](https://user-images.githubusercontent.com/1355810/163568361-853769b5-0769-43be-b4ff-a88d2a7c83c1.png)
_Filtering by name and a warning for duplicate binding_

![image](https://user-images.githubusercontent.com/1355810/163568372-061644c9-cdd9-46d5-8dca-8f72a86771a0.png)
_No matching hotkeys_

Closes #19898

dragunoff committed:
Add hotkey filtering functionality (by name and by context)

pchote commented:
How about:

```diff
diff --git a/OpenRA.Mods.Common/Widgets/Logic/Settings/HotkeysSettingsLogic.cs b/OpenRA.Mods.Common/Widgets/Logic/Settings/HotkeysSettingsLogic.cs
index 4a078409bc..8385cd2cc2 100644
--- a/OpenRA.Mods.Common/Widgets/Logic/Settings/HotkeysSettingsLogic.cs
+++ b/OpenRA.Mods.Common/Widgets/Logic/Settings/HotkeysSettingsLogic.cs
@@ -216,7 +216,7 @@ void InitHotkeyRemapDialog(Widget panel)
                        var duplicateNotice = panel.Get<LabelWidget>("DUPLICATE_NOTICE");
                        duplicateNotice.TextColor = ChromeMetrics.Get<Color>("NoticeErrorColor");
                        duplicateNotice.IsVisible = () => !isHotkeyValid;
-                       var duplicateNoticeText = new CachedTransform<HotkeyDefinition, string>(hd => hd != null ? duplicateNotice.Text.F(hd.Description) : duplicateNotice.Text);
+                       var duplicateNoticeText = new CachedTransform<HotkeyDefinition, string>(hd => hd != null ? duplicateNotice.Text.F(hd.Description, hd.Contexts.First(c => selectedHotkeyDefinition.Contexts.Contains(c))) : duplicateNotice.Text);
                        duplicateNotice.GetText = () => duplicateNoticeText.Update(duplicateHotkeyDefinition);
 
                        var originalNotice = panel.Get<LabelWidget>("ORIGINAL_NOTICE");
diff --git a/mods/cnc/chrome/settings-hotkeys.yaml b/mods/cnc/chrome/settings-hotkeys.yaml
index 310956c438..f5eee56fdf 100644
--- a/mods/cnc/chrome/settings-hotkeys.yaml
+++ b/mods/cnc/chrome/settings-hotkeys.yaml
@@ -128,7 +128,7 @@ Container@HOTKEYS_PANEL:
                                                                        Width: PARENT_RIGHT
                                                                        Height: PARENT_BOTTOM
                                                                        Font: Tiny
-                                                                       Text: This is already used for "{0}" in an overlapping context
+                                                                       Text: This is already used for "{0}" in the {1} context
                                                Button@OVERRIDE_HOTKEY_BUTTON:
                                                        X: PARENT_RIGHT - 3 * WIDTH - 30
                                                        Y: 20
```
(plus the other yaml files)?

dragunoff commented:
Rebased and implemented https://github.com/OpenRA/OpenRA/pull/19979#issuecomment-1107660411:
![image](https://user-images.githubusercontent.com/1355810/165170445-b8eb98a0-0469-448e-acda-27b00b647589.png)

abcdefg30 commented on OpenRA.Mods.Common/Widgets/Logic/Settings/HotkeysSettingsLogic.cs:
```diff
@@ -231,5 +301,48 @@ void OverrideHotkey()
+		bool IsHotkeyVisibleInFilter(HotkeyDefinition hd)
+		{
+			var filter = filterInput.Text;
+			var isFilteredByName = false;
+			var isFilteredByContext = false;
+
+			if (string.IsNullOrWhiteSpace(filter) || hd.Description.Contains(filter, StringComparison.OrdinalIgnoreCase))
+				isFilteredByName = true;
+
+			if (currentContext == "Any" || hd.Contexts.Contains(currentContext))
+				isFilteredByContext = true;
+
+			return isFilteredByName && isFilteredByContext;
```
```suggestion
			var isFilteredByName = string.IsNullOrWhiteSpace(filter) || hd.Description.Contains(filter, StringComparison.OrdinalIgnoreCase);
			var isFilteredByContext = currentContext == "Any" || hd.Contexts.Contains(currentContext);
			return isFilteredByName && isFilteredByContext;
```

abcdefg30 commented on OpenRA.Mods.Common/Widgets/Logic/Settings/HotkeysSettingsLogic.cs:
```diff
@@ -144,21 +165,67 @@ Action ResetPanel(Widget panel)
+			var duplicateNoticeText = new CachedTransform<HotkeyDefinition, string>(hd =>
+				hd != null ?
+				duplicateNotice.Text.F(hd.Description, hd.Contexts.First(c => selectedHotkeyDefinition.Contexts.Contains(c))) :
+				duplicateNotice.Text);
```
The duplicate notice has `{0}` and `{1}` in it, so does it even make sense to display that here without `.F`? Also, if `hd` is `null` in this context, then `duplicateHotkeyDefinition` is also `null`, correct? In that case this text is not displayed anyway, so I'd suggest putting an empty string.

dragunoff commented:
Update: rebased and applied the fixups suggested by @abcdefg30

dragunoff committed:
Remove unneeded yaml nodes from hotkey settings

dragunoff commented:
Update: Slight scope creep - I'm using the opportunity to remove a couple of stray yaml nodes that are no longer needed.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/fc0a11aa15f9ff2219de3346bdd66754863fd122...3d080c5fc2178159d0771b742a009d41afecf1c5)
